### ISSUE TITLE
[Feature Request] Prevent Orchestrations loading large Activity Results into memory when they are not used in the Orchestration directly.

### ISSUE THREAD

**Issue by olitomlinson** [Azure/azure-functions-durable-extension#1682]
**Is your feature request related to a problem? Please describe.**

Not a problem per-se, just an optimisation.

**Describe the solution you'd like**

Given an Orchestration where multiple Activities are called
And each Activity returns a large Object graph to the orchestration
When all Activities are complete, references to the Object graphs are returned to the orchestration (rather than the actual Object Graphs themselves)
Then the list of Object references are passed as an input to a subsequent Activity
And in the Activity, the binding automatically loads the reference/s and deserializes into a List of the Object Graphs.

**Describe alternatives you've considered**

I could manually use blob storage to store the object graphs and pass around the references, but this is just unsightly boilerplate. It would be very convenient if the extension could do this for me.

**Additional context**

The purpose of this is stop the Orchestration having to load and unload many large objects (from the result of an activity) into memory, just for them to passed to subsequent Activities.

**Comment by cgillum** [comment]
Hey @olitomlinson, can I think of this feature request as being similar to the [claim check pattern](https://docs.microsoft.com/en-us/azure/architecture/patterns/claim-check)?  Basically, you want the ability to pass around _references_ to your large object graphs  without actually pulling them into and out of the orchestration?

If so, I agree this is something we should consider.

**Comment by olitomlinson** [comment]
@cgillum I never knew there was a name for it it, but yes, that claim-check pattern sounds right! 

If the result of an Activity returns a large object and I know I’m not going to make use of the large object directly in the orchestration, other than to forward that object to a subsequent activity, why bother loading it? Just pass the reference 🙂

**Comment by ConnorMcMahon** [comment]
I think whatever approach we end up taking will have to be explicitly opted into., The reason for this is that we can't just lazily load activity results once the orchestration has started, as we can't perform IO at that point. Therefore, we have to know what data we need to fully load before starting the orchestration, and I think the easiest way to achieve that is via some opt-in flag and/or datatype.

**Comment by olitomlinson** [comment]
@ConnorMcMahon 

I agree. Opt-in should be the way, and not the default experience.

I don't know if this is viable, but I had in my mind something like `Reference<T>` class which could instruct the extension to automatically store the blob into storage, and then also be used to tell the extension not to load `T` in the orchestration.

```c#
[FunctionName("Orchestration")]
public async Task Orchestration([OrchestrationTrigger] IDurableOrchestrationContext context)
{
    List<Reference<Foo>> fooRefs = new List<Reference<Foo>>();        
    foreach(var i in Enumerable.Range(0,100)
    {
        fooRefs.Add(await context.CallActivityAsync<Reference<Foo>>("GetFoo", 1));
    }

    await context.CallActivityAsync("ProcessFoos", fooRefs, LazyLoadReferences : true | false );
}

[FunctionName("GetFoo")]
public async Task<Reference<Foo>> GetFoo([ActivityTrigger] int id)
{
    var foo = await getLargeFooObject();
    return new Reference<Foo>(foo);
}
        
[FunctionName("ProcessFoo")]
public async Task ProcessFoo([ActivityTrigger] List<Reference<Foo>> foo)
{
    foreach(var ref in foo) 
    {
        Foo = await ref.GetDataAsync();
    }
}
```

**Comment by ConnorMcMahon** [comment]
Yeah, I agree a data type like `Reference<T>` would be a great way to support this in C#, though we also want a solution for languages like Node and Python. I suppose we could define similar data types in those languages, we would just need to make sure that we can safely mark those as "references" back in the extension without accidentally doing so for any existing types in those languages.

**Comment by ConnorMcMahon** [comment]
Starting on the design for implementing this. Hopefully it will make it into our v2.5.0 release.

**Comment by olitomlinson** [comment]
@ConnorMcMahon 

Great news! I have a few real-world use cases for it so looking forward to it!

**Comment by ConnorMcMahon** [comment]
We are still working on the design of this feature, as it is quite complicated to implement across backends and all of our languages. We have made good progress, but there are some edge cases we are still working out.

**Comment by hctan** [comment]
Hi, regarding this feature, i came across this [link](https://stackoverflow.com/questions/59167994/maximum-activity-response-data-size-limit-in-azure-durable-functions) which mentioned that durable function would automatically store the activity output payload in blob if its > 64kb. If this is the case, the alternative mentioned in the feature - storing into blob and pass the blob reference, would it help? It looks like doing the same thing, no?

We are also having this issue where our activity can sometimes receives huge response (over 100mb) when retrieving items from sharepoint endpoint, and we are passing the payload to subsequent activity. The memory usage then jumps up (when there's many concurrent requests with large payload) and there's orchestration that get stuck in running state.

**Comment by cgillum** [comment]
@hctan the problem with the automatic storage into blobs is that we must also automatically deserialize those large blobs when loading the orchestration into memory since we can't know whether or not the orchestration needs to reference that data. Blob references would be a contract indicating that the orchestration doesn't need to access the blob data and can therefore skip loading it, which would result in significant performance gains.
